SeQuant/core/index.hpp

```diff
@@ -79,62 +110,173 @@ class Index : public Taggable {
 
   Index() = default;
 
-  /// @param label the label, does not need to be unique
+  /// copy constructor
+  /// @warning memoized data (label, full_label) is not copied
+  Index(const Index &idx) noexcept
+      : Taggable(idx),
+        space_(idx.space_),
+        ordinal_(idx.ordinal_),
+        proto_indices_(idx.proto_indices_),
+        symmetric_proto_indices_(idx.symmetric_proto_indices_) {}
```

Comment:
I think I would implement the copy ctor in terms of the copy-assignment ctor in order to reduce code duplication.
```suggestion
  Index(const Index &idx) noexcept { *this = idx; }
```

Reply:
I don't know if the compiler will make this as efficient as the explicit version. For core concepts like Index I don't mind being verbose

Reply:
I would assume that the code will be equally efficient. But alright :+1:

Reply:
I would certainly hope so, but strictly speaking default initialization will occur first, then the data will be overwritten by copy assignment.